src/middlewared/middlewared/plugins/auth.py

```diff
@@ -325,9 +339,130 @@ def token(self, app, token):
         return True
 
 
+class TwoFactorAuthService(ConfigService):
+
+    class Config:
+        datastore = 'system.twofactorauthentication'
+        datastore_extend = 'auth.twofactor.two_factor_extend'
+        namespace = 'auth.twofactor'
+
+    @private
+    def two_factor_extend(self, data):
+        data['secret'] = self.middleware.call_sync('pwenc.decrypt', data['secret'])
+
+        for srv in ['ssh']:
+            if not data['services'].get(srv):
+                data['services'][srv] = False
+
+        data['secret_hex'] = base64.b16encode(base64.b32decode(data['secret'])).decode()
```

Comment:
We already have secret, why do we need secret_hex in the service? If its only used in the config file, perhaps it can be encoded there?